File: Java/programmers/문자열압축.java
package java. java.programmers;
import java.util.*;

public class 문자열압축 {

	class Solution {
	    public int solution(String s) {
	        int answer = s.length();
	       
	        for(int i=1; i<=s.length()/2; i++){
	            String temp = "";
	            String comp = "";
	            int cnt = 1;
	            for(int j =0; j<s.length(); j+=i){
	                if(j+i<=s.length()){
	                    if(j==0) {
	                        temp = s.substring(j,j+i);
	                        continue;
	                    }
	                    if(temp.equals(s.substring(j,j+i))){
	                        cnt++;
	                    }else{
	                        if(cnt ==1 ){
	                            comp+= temp;
	                        }else{
	                            comp+= Integer.toString(cnt)+ temp;
	                            cnt =1;
	                        }
	                      temp = s.substring(j,j+i);
	                    }
	                }
	                
	                else if(j+i >=s.length()){
	                    comp+=s.substring(j,s.length());
	                }
	         
	            }
	            
	            if(cnt ==1 ){
	                comp+= temp;
	            }else{
	                comp+= Integer.toString(cnt)+ temp;
	            }
	           if(answer > comp.length()){
	               answer = comp.length();
	           }
	            
	            System.out.println(comp);
	        }
	       
	        
	        return answer;
	    }
	}
}

File: JavaScript/programmers/문자열압축.js
function solution(s) {
  let answer = s.length;
  let half = Math.floor(s.length / 2);
  // 등호가 필요한 이유는 slice 마지막은 잘리지 않음
  for (let i = 1; i <= half; i++) {
    let temp = s.slice(0, i); // 초기 비교값
    let comp = ""; // 압축 문자열 저장
    let cnt = 1; // 압축 개수를 저장하는 변수
    // 처음부터 개수 크기만큼 확인
    for (let j = i; j < s.length; j += i) {
      if (temp === s.slice(j, j + i)) {
        cnt++; // 비교 문자열과 같다면 카운터 증가 후 다음으로
      } else {
        // 다르다면 압축 문자열에 저장 -> 1 이상이면 압축이므로 cnt를 붙여줌
        if (cnt > 1) comp += cnt + temp;
        else comp += temp;
        temp = s.slice(j, j + i);
        cnt = 1;
      }
    }

    if (cnt > 1) comp += cnt + temp;
    else comp += temp;

    answer = Math.min(answer, comp.length);
  }

  return answer;
}
